spec.md

```diff
@@ -336,19 +336,30 @@ on the definition of OPTIONAL.
   event format defines the relationship in
   [section 3.1](./json-format.md#31-special-handling-of-the-data-attribute).
 
-  When this attribute is omitted, the `data` attribute simply follows the event
-  format's encoding rules. For the JSON event format, the `data` attribute value
-  can therefore be a JSON object, array, or value.
-
   For the binary mode of some of the CloudEvents transport bindings, where the
   `data` content is immediately mapped into the payload of the transport frame,
   this field is directly mapped to the respective transport or application
   protocol's content-type metadata property. Normative rules for the binary mode
   and the content-type metadata mapping can be found in the respective transport
   mapping specifications.
 
+  Implementations MAY introduce this parameter when encoding `Integer`, `Map`
```

Comment:
I would turn that on its head as a MUST rule for Binary/String. This should also be "attribute", not "parameter":

> The attribute MUST be set for `Binary` content in the `data` field, with an accurate media-type.
> If no IANA registered media-type is available, a vendor extension or `application/octet-stream`
> MAY be used.
>
> The attribute SHOULD be set for `String` content in the `data` field. If set, the media-type MUST     
> describe a text format. If the attribute is omitted for `String` content, the default is `text/plain`.
> If the character encoding is not UTF-8 or one of its subsets (US-ASCII, ISO-8859-1) or if the 
> character encoding of the overall event diverges, the character encoding MUST be declared with
> a charset parameter and the text MUST be encoded as `Binary`.
>
> The attribute MAY be set for all other types in order to enforce encoding of the `data` attribute 
> according to the rules of a particular event format, which might differ from the event format 
> chosen for the overall event. Using "application/json" will encode the `data` content according to 
> the JSON event format rules. When the event itself is JSON encoded, the content is inlined into 
> the JSON object's `data` field. When the event is encoded in a different event format, the content 
> is carried as `String` or `Binary` under consideration of the aforementioned character encoding rules.
>
> If the attribute is omitted, the content of the `data` attribute is encoded according to the rules of
> the chosen event format, which also allows for transcoding of the content as an event is 
> forwarded through routes that use different event formats. An event might initially published via
> MQTT with Protobuf encoding for preserving bandwidth, but then routed onwards via Kafka 
> using JSON encoding since the consumers are expecting JSON content.  
>
> For using the "binary" mode of any transport binding, the `datacontenttype` MUST be declared.

Reply:
@clemensv 
If an event router receives an event in HTTP binary mode, how can it distinguish events for which the producer explicitly set datacontenttype to application/json from an event that was produced without a datacontenttype? How does it know when to transcode the content of the data attribute and when to continue treating it just as a JSON document?

Reply:
In binary mode... does it matter? In binary mode, I believe that we can only really be concerned about how to get the CE specific attributes, which means the `ce-` http headers. The http body is (basically) out of scope for us (aside from the abstract notion of it mapping to our `data` attribute) but whatever normal processing the receiver has should apply, just as if CloudEvents were never used at all.  No?

Reply:
Perhaps I am still missing something, but isn't there a difference between data adhering to the CE type system, that is only represented as JSON for a single hop (HTTP binary), and data that was intended to be a JSON document? I don't see, how it is possible to distinguish these cases after an HTTP binary hop.

Reply:
Maybe I'm missing something too :-)  But to me, `data` should have zero restrictions on it regardless of binary vs structured. CE should never restrict, or change, what can be sent for business logic (ignoring serialization aspects). The fact that we are even having this discussion kind of bothers me and I think it's due to presence of our type system. I don't think it should apply to `data` since doing so implies some kind of restrictions and I hope that wasn't the intent.

Using it for our other attributes just seems unnecessary to me since we require a string mapping and unknown attribute would remain as strings. Known attributes, since they're known, can be converted from string to **whatever** by the entity that understands it's type.  And that can all be done (and specified in the spec) w/o a formal type system that seems to be introducing more questions than necessary.

To me, if we defined a new attribute called `size`, then:
```
name: size
type: int
constraints: must be between 0 and 100
```
can be expressed as:
```
name: size
constraints: must be a string representation of an integer between 0 and 100
```
and I think we'd get the exact same serializations and code written, but less questions/concerns.  And even with this, if the transport does support types metadata, the spec for that transport can specify that an `int` type be used if it wants.